Documentation for OQL (#8431)
mendix/docs: content/en/docs/refguide/modeling/domain-model/oql/oql-clauses.md

In addition, OQL can use predefined relations (associations) to easily join objects without having to calculate which columns should be coupled. Despite these differences, many SQL keywords also work in OQL.

These are some examples of OQL queries:

* `SELECT LastName FROM Sales.Customer` – retrieves the names of all customers
* `SELECT FirstName FROM Sales.Customer WHERE LastName = 'Jansen'` – retrieves the first name of all customers with name "Jansen"
* `SELECT SUM(TotalAmount) FROM Sales."Order" WHERE IsPaid = true` – retrieves the sum of the total prices of all paid orders (`Order` needs to be wrapped in quotes, see the [Reserved Words](#reserved-oql-words) section below)

{{% alert color="info" %}}
OQL queries do not take security into account out-of-the-box. This means that you can use OQL to manually define custom security expressions. In some cases, handling security yourself using OQL—instead of using the out-of-the-box security of XPath—may result in faster queries.
{{% /alert %}}

Try your OQL example online with the [OQL Playground](https://service.mendixcloud.com/p/OQL) demo app. 

## OQL Clauses

> 

Reviewer: I think this is more appropriate for the OQL landing page. So we have the structure:

    OQL Landing Page (what is OQL where do you use it, how to use it - or links to pages on how to use it
        OQL Clauses
        OQL Expressions
        (other OQL docs?)

I've put a dummy OQL Landing page there for now - we can use some of what is in the existing OQL page, but I think we should ideally have a completely new page?

OQL clauses are reserved words that structure an OQL query into sections. Each section, and therefore clause, is responsible for a different aspect of the query. Every query has 2 mandatory clauses: `SELECT` and `FROM`. For example:

```sql
SELECT LastName, Address
FROM Sales.Customer
```

These basic clauses define which data needs to be retrieved and which entity or entities should be used as a source. Other clauses are not mandatory. They are used to specify limitations and other rules on the data being retrieved.

> Other clauses are not mandatory, but if used, they should be present in a fixed order:

Reviewer: ```suggestion
These basic clauses define the data that needs to be retrieved and which entity or entities should be used as a source. Other clauses are not mandatory. They are used to specify limitations and other rules on the data being retrieved. If they are used, they should be present in the following order:
```

1. `SELECT`
2. `FROM`
3. `WHERE`
4. `GROUP BY`
5. `HAVING` (allowed only in combination with `GROUP BY`)
6. `ORDER BY`
7. `LIMIT`
8. `OFFSET`

### `SELECT` clause

> The `SELECT` clause specifies which entity attributes or other specified data must be retrieved. The `SELECT` clause consists of the term `SELECT` and one or more expressions. Each expression must be separated by a comma. Each expression defines a column in the result. Each expression can have an alias, which will be the name of the column in the result.

Reviewer: You say that you have `SELECT` plus expressions. However, you explain these expressions in this document, not in the **OQL Expressions** document. I think this is a bit confusing - is there another word we can use (sub-clauses, or definitions?) instead of **expressions** here?

Also, you say that _Each expression defines a column in the result._ However, the `*` expression defines multiple columns in the result, doesn't it?

#### Syntax

The syntax is as follows:

```sql
SELECT [ DISTINCT ]
	{
			*
		| { entity_name | from_alias }.*
		| { expression [ [ AS ] column_alias ] } [ ,...n ]
	}
```

#### Selecting Particular Attributes

A simple way to define what is to be retrieved is to specify particular attributes of an entity. For example:

```sql
SELECT FirstName AS FName, LastName AS LName
FROM Sales.Customer
```

It is possible to specify aliases for attributes: `FirstName AS FName`. The keyword `AS` is not mandatory, so the following query is equivalent to the one above:

```sql
SELECT FirstName FName, LastName LName
FROM Sales.Customer
```

Alias is an alternative name to replace the column name in the result. When the name attribute is retrieved, the result column is "Name". With an alias, you can specify another result column name, like "Customer_Name". An alias cannot contain spaces.

{{% alert color="info" %}}

> In some scenarios, aliases in `FROM` are mandatory. Aliases are mandatory when defining a view entity.

Reviewer: Is this the only case of a mandatory use of an alias, or just an example?
If there are more, then can we define them all?

{{% /alert %}}

#### Selecting all attributes using `*`

`*` (asterisk) specifies that all attributes from all entities in the `FROM` clause should be returned. `entity_name/*` and `from_alias/*` specify that all attributes of the specified entity or expression of the `FROM` clause should be returned. `entity_name` can be optionally put in double quotes. If the entity name is a [reserved OQL word](/refguide/oql/#reserved-oql-words) (like `Order` or `Group`), double quotes are mandatory.

> ##### Examples

Reviewer: Do we need four examples. Can we just have two examples: the simplest `SELECT *` and the most complex?

Author: I reduced the number of examples

This query returns all attributes of `Sales.Customer`

```sql
SELECT *
FROM Sales.Customer
```

This query returns all attributes of objects of `Sales.Customer` entity all attributes of associated objects of `Sales.Request` (see TODO)

```sql
SELECT *
FROM Sales.Customer

> JOIN Sales.Customer/Sales.Customer_Request/Sales.Request

Reviewer: Should we have an example of JOIN before we have documented it? I think this would be better in the **JOIN** section.

Author: I removed this example, but other ones are used to illustrate `TableName/*`

```

This query returns all attributes of objects of `Sales.Request` that are associated to `Sales.Customer` objects (see TODO)

```sql
SELECT Sales.Request/*
FROM Sales.Customer
JOIN Sales.Customer/Sales.Customer_Request/Sales.Request

> ```

Reviewer: Using SQL for the language here has a weird effect on the markdown formatting as `/*` is not part of the SQL standard. We may have to not specify a language on these code blocks.
![image](https://github.com/user-attachments/assets/ad16a583-b347-4dac-b7b5-ed81190c3148)

```sql
SELECT DISTINCT *
FROM Sales.Customer
```

| ID              | FName | LName |
| --------------- | ----- | ----- |
| 562949953421521 | John  | Doe   |
| 562949953421683 | Jane  | Doe   |
| 562949953421777 | Jane  | Doe   |
| 562949953421923 | Jane  | Moose |

If used, `DISTINCT` should follow directly after `SELECT`. It is not possible to request only some attributes to be distinct.

#### Expressions

It is possible to use more complex expressions in `SELECT`. That is explained in detail in [Expressions](TODO). Here are some examples:

It is possible to use a function of an attribute. For instance:

```sql
SELECT LastName Name, LENGTH(LastName) NameLength
FROM Sales.Customer
```

| Name  | NameLength |
| ----- | ---------- |
| Doe   | 3          |
| Doe   | 3          |
| Doe   | 3          |
| Moose | 5          |

It is possible to use an aggregate function:

```sql
SELECT COUNT(*) AS CustomerCount
FROM Sales.Customer
```

| CustomerCount |
| ----------    |

> | 5             |

Reviewer: I think these should be in the expressions doc. We don't need them here.

> Another possibility is subqueries. A subquery can refer to other tables and expressions in `FROM`.

Reviewer: We currently have a whole document on subqueries (https://docs.mendix.com/refguide/oql-subqueries/). I think we should just move that into the OQL section of the documentation and refer to that rather than having examples here?

`entity_name` can optionally be put in double quotes. If the entity name is a [reserved OQL word](/refguide/oql/#reserved-oql-words) (like `Order` or `Group`), double quotes are mandatory.

{{% alert color="warning" %}}
Specifying all attributes will also return attributes which are normally hidden in the Domain Model, such as the `ID` of each object.
{{% /alert %}}

##### Examples

This query returns the values of all the attributes of `Sales.Customer`

```sql
SELECT *
FROM Sales.Customer
```

This query returns all the attributes of objects of `Sales.Customer` entity all attributes of associated objects of `Sales.Request` (see TODO)

```sql
SELECT *
FROM Sales.Customer
JOIN Sales.Customer/Sales.Customer_Request/Sales.Request
```

This query returns all attributes of objects of `Sales.Request` that are associated to `Sales.Customer` objects (see TODO)

```sql
SELECT Sales.Request/*
FROM Sales.Customer
JOIN Sales.Customer/Sales.Customer_Request/Sales.Request
```

This query is equivalent to the previous one, but it uses table aliases

```sql
SELECT Req/*
FROM Sales.Customer Cust
JOIN Cust/Sales.Customer_Request/Sales.Request Req
```

> #### Selecting Distinct Values with `DISTINCT`

Reviewer 2: This section is being referenced as #Distinct by oql-expressions

Reviewer 2: This can be resolved

```sql
SELECT DISTINCT LastName LName
FROM Sales.Customer
```

returns

| LName |
| ----- |
| Doe   |
| Moose |

If multiple attributes are selected, the result is all the unique combinations that are present in the database:

```sql
SELECT DISTINCT FirstName FName, LastName LName
FROM Sales.Customer
```

returns

| FName | LName |
| ----- | ----- |
| John  | Doe   |
| Jane  | Doe   |
| Jane  | Moose |

`DISTINCT` can also be combined with `*`.

{{% alert color="warning" %}}
If you specify an entity name in `FROM`, this will return all columns, including the unique column `ID`. This means that adding `DISTINCT` does not affect the result.

In more complex cases, `SELECT DISTINCT *` can become useful.
{{% /alert %}}

```sql
SELECT DISTINCT *
FROM Sales.Customer
```

returns

| ID              | FName | LName |
| --------------- | ----- | ----- |
| 562949953421521 | John  | Doe   |
| 562949953421683 | Jane  | Doe   |
| 562949953421777 | Jane  | Doe   |
| 562949953421923 | Jane  | Moose |

#### Expressions

It is possible to use more complex expressions in `SELECT`. That is explained in detail in [OQL Expressions](/refguide/oql-expressions). Here are some examples:

> It is possible to use a function of an attribute. For instance:

Reviewer: Not sure about this. I initially didn't like the word "function" but I see that SQL has a lot of functions and that we are using it here in a similar way.
But the function isn't a function of the attribute (unlike a method in OOP) so perhaps 
```suggestion
It is possible to apply a function to an attribute. For instance:
```

Please note that if you use a subquery as an expression in `FROM`, then such subquery should always return  a single column, and the number of rows should be at most one. If the subquery returns more than one row or a number of columns different than one, that will lead to an exception during runtime.
{{% /alert %}}

### `FROM` clause

The `FROM` clause specifies the entities or other source from which the data must be retrieved. This clause starts with the `FROM` keyword, followed by an entity name. To select data from other entities, add these entities via the `JOIN` keywords. This syntax is a little more strict than the official `SQL FROM` clause syntax.

#### Syntax

This is an example of the full syntax:

```sql
FROM
	{
		entity_name | ( sub_oql_query )
	}
	[ [ AS ] from_alias ]

	{
		{ INNER | { { LEFT | RIGHT | FULL } [ OUTER ] } } JOIN
		entity_path | entity_name | ( sub_oql_query ) [ [ AS ] from_alias ]
		[ ON <constraint> ]
	} [ ,...n ]
```

#### `SELECT` from one entity

An example of a simple select from one entity:

```sql
SELECT Sales.Customer/LastName
FROM Sales.Customer
```

An alias can be used to replace entity name:

```sql
SELECT Cust/LastName

> FROM Sales.Customer Cust

Reviewer: Would this be clearer if we included `AS`?
```suggestion
FROM Sales.Customer AS Cust
```

| 562949953421521   | John      | Doe      | 1688849860264654 | Caribou      | -1     |
| 562949953421923   | Jane      | Moose    | 1688849860264654 | Caribou      | -1     |
| 562949953422131   | Jim       | Elk      | 1688849860264654 | Caribou      | -1     |

It is possible to specify conditions in `WHERE` (see more details and examples in [`WHERE` clause](TODO))

```sql
SELECT *
FROM Sales.Customer Cust, Sales.Request Req
WHERE Cust.LastName = Req.CustomerName
```
| sales$customer.ID | FirstName | LastName | sales$request.ID | CustomerName | Number |
| ---------------   | -----     | -----    | ---------------- | ------------ | ------ |
| 562949953421521   | John      | Doe      | 1688849860264073 | Doe          | 1      |
| 562949953421923   | Jane      | Moose    | 1688849860264231 | Moose        | 2      |

It is possible to specify columns in `SELECT`. To avoid ambiguity in case opf duplicate attribute names, it is possible to specify entity name or alias. Respectively, `<module>.<entity>/<attribute>` or `<alias>/<attribute>`. It is possible to retrieve alll attributes of a particular entity using `<alias>/*`

> ```sql

Reviewer: I think we will have to leave the syntax highlighting off this as the `*` messes up the formatting and it looks weird (https://meta.stackoverflow.com/questions/401610/code-block-containing-asterisks-displaying-in-italics).
So
```suggestion
```
```

```
SELECT Cust/FirstName, Req/*
FROM Sales.Customer Cust, Sales.Request Req
WHERE Cust.LastName = Req.CustomerName
```

returns

| FirstName | sales$request.ID | CustomerName | Number |
| -----     | ---------------- | ------------ | ------ |
| John      | 1688849860264073 | Doe          | 1      |
| Jane      | 1688849860264231 | Moose        | 2      |

#### Select from Multiple Tables using `JOIN` {#join}

OQL supports the `JOIN` syntax, which is similar to SQL. There are 4 main types of `JOIN`:

* `INNER JOIN` or simply `JOIN`
* `LEFT JOIN` or `LEFT OUTER JOIN`
* `RIGHT JOIN` or `RIGHT OUTER JOIN`
* `FULL JOIN` or `FULL OUTER JOIN`

In addition to standard SQL syntax, OQL supports joins over associations.

The syntax is as follows:

```sql
{ INNER | { { LEFT | RIGHT | FULL } [ OUTER ] } } JOIN
entity_path | entity_name | ( sub_oql_query ) [ [ AS ] from_alias ]
[ ON <constraint> ]
```

##### entity_path

`entity_path` specifies the entity to join and the path from an earlier defined entity in the `FROM` clause to this entity.

> The example path `Crm.Customer/Crm.Customer_Address/Crm.Address` defines an association path from the entity **Crm.Customer** to a new entity **Crm.Address**.

Reviewer: ```suggestion
The example path `Crm.Customer/Crm.Customer_Address/Crm.Address` defines an association path from the entity **Crm.Customer** to an associated entity **Crm.Address**.
```

| Veidt  | Rotterdam | 23    | Address 2 | 2              |
| Veidt  | Rotterdam | 1     | Address 6 | 6              |
| Veidt  | Utrecht   | 2     | Address 5 | 5              |

The following query retrieves total stock per brand:

```sql
SELECT Brand, SUM(Stock) AS SumStock
FROM Sales.Location
GROUP BY Brand
```

| Brand  | SumStock |
| ------ | -----    |
| Cinco  | 5        |
| Rekall | 12       |
| Veidt  | 26       |

It is possible to specify multiple aggregate functions in combinantion with `GROUP BY`:

```sql
SELECT
	Brand,
	SUM(Stock) AS SumStock,
	MIN(Stock) AS MinStock,
	MAX(Stock) AS MaxStock
FROM Sales.Location
GROUP BY Brand
```

| Brand  | SumStock | MinStock | MaxStock |
| ------ | -----    | -----    | -----    |
| Cinco  | 5        | 5        | 5        |
| Rekall | 12       | 3        | 9        |
| Veidt  | 26       | 1        | 23       |

It is possible to group by multiple attributes:

```sql
SELECT Brand, City, SUM(Stock) AS SumStock
FROM Sales.Location
GROUP BY Brand, City
```

| Brand  | SumStock |
| ------ | -----    |
| Cinco  | 5        |
| Rekall | 9        |
| Rekall | 3        |
| Veidt  | 24       |

> | Veidt  | 2        |

Reviewer: Why does this table not include the `City` column?

ORDER BY LocationNumber
LIMIT 3
```

| Brand  | City      | LocationNumber |
| ------ | -----     | -----          |
| Cinco  | Rotterdam | 1              |
| Veidt  | Rotterdam | 2              |
| Rekall | Zwolle    | 3              |

#### `OFFSET` clause

`OFFSET` specifies how many rows must be skipped before returning the result rows.

For example, the following query retrieves all records except the first 2:

```sql
SELECT Brand, City, LocationNumber
FROM Sales.Location
ORDER BY LocationNumber
OFFSET 2
```

| Brand  | City      | LocationNumber |
| ------ | -----     | -----          |
| Rekall | Zwolle    | 3              |
| Rekall | Utrecht   | 4              |
| Veidt  | Utrecht   | 5              |
| Veidt  | Rotterdam | 6              |

`LIMIT` and `OFFSET` can be combined in one query.

For example, the following query skips 2 records and then returns 3 records:

```sql
SELECT Brand, City, LocationNumber
FROM Sales.Location
ORDER BY LocationNumber
LIMIT 3
OFFSET 2
```

| Brand  | City      | LocationNumber |
| ------ | -----     | -----          |
| Rekall | Zwolle    | 3              |
| Rekall | Utrecht   | 4              |
| Veidt  | Utrecht   | 5              |

> ### Subqueries

Reviewer: Will this replace the current [OQL Subqueries page](https://docs.mendix.com/refguide/oql-subqueries/)?

That page has subquery in HAVING. Are we deliberately not documenting this because subqueries in GROUP BY are database dependent?

Author: GROUP by is database dependent, but HAVING should work in every database. I'm pushing some examples for that.

---
title: "OQL Clauses"
url: /refguide/oql-clauses/
weight: 10
---

## OQL Clauses

OQL clauses are reserved words that structure an OQL query into sections. Each section, and therefore clause, is responsible for a different aspect of the query. Every query has 2 mandatory clauses: `SELECT` and `FROM`. For example:

```sql
SELECT LastName, Address
FROM Sales.Customer
```

These basic clauses define the data that needs to be retrieved and which entity or entities should be used as a source. Other clauses are not mandatory. They are used to specify limitations and other rules on the data being retrieved. If they are used, they should be present in the following order:

1. `SELECT`
2. `FROM`
3. `WHERE`
4. `GROUP BY`
5. `HAVING` (allowed only in combination with `GROUP BY`)
6. `ORDER BY`
7. `LIMIT`
8. `OFFSET`

### `SELECT` clause

The `SELECT` clause specifies which entity attributes or other specified data must be retrieved. The clause returns all the values of objects which match the `SELECT` clause.

> The `SELECT` clause consists of the term `SELECT` and one or more column definitions. Each column definition must be separated by a comma. Each column definition defines a column or a set of columns in the result. Each single value column definition can have an alias, which will be the name of the column in the result.

Reviewer: I think we need to say somewhere that a "column" is the same as an "attribute"?

```sql
SELECT FirstName FName, LastName LName
FROM Sales.Customer
```

An alias is an alternative name which replaces the column name in the result. For example, when the `name` attribute is retrieved, the result column is "Name". With an alias, you can call the result column something else, like "Customer_Name". An alias cannot contain spaces.

{{% alert color="info" %}}
Aliases are mandatory when defining a view entity.
{{% /alert %}}

#### Selecting All Attributes Using `*`

Using `*` (asterisk) in the SELECT clause specifies that the values of all attributes from all entities in the `FROM` clause should be returned.

Specifying `entity_name/*` and `from_alias/*` specify that the values of all attributes of the specified entity or expression of the `FROM` clause should be returned.

`entity_name` can optionally be put in double quotes. If the entity name is a [reserved OQL word](/refguide/oql-clauses/#reserved-oql-words) (like `Order` or `Group`), double quotes are mandatory.

{{% alert color="warning" %}}

> Specifying all attributes will also return attributes which are normally hidden in the Domain Model, such as the `ID` of each object.

Reviewer: Is this a warning or just info?

The following query returns only aggregate rows for brands with more than one location:

```sql
SELECT Brand, SUM(Stock) AS SumStock, COUNT(*) AS LocationCount
FROM Sales.Location
GROUP BY Brand
HAVING COUNT(*) > 1

```

| Brand  | SumStock | LocationCount |
| ------ | -----    | ----          |
| Rekall | 12       | 2             |
| Veidt  | 26       | 3             |

You can use aggregate functions that are not present in `SELECT`:

```sql
SELECT Brand
FROM Sales.Location
GROUP BY Brand
HAVING COUNT(*) > 1 AND SUM(Stock) < 20

```

| Brand  |
| ------ |
| Rekall |

You can also use more complex expressions such as subqueries and functions in `HAVING`:

```sql
SELECT Brand
FROM Sales.Location loc
GROUP BY Brand
HAVING
	COUNT(*) > 1
	AND
	EXISTS (
		SELECT *
		FROM Sales.Storage stor
		WHERE stor/AvailableStorage >= SUM(loc/Stock)
	)
```

### `ORDER BY` clause

The `ORDER BY` clause specifies the sort order used on columns returned in a `SELECT` statement. Multiple columns can be specified. Columns are ordered in the sequence of the items in the `ORDER BY` clause.

{{% alert color="info" %}}

> This clause can include items that do not appear in the `SELECT` clause, except when `SELECT DISTINCT` is specified or when a `GROUP BY` clause exists. When `UNION` is used, the column names or aliases must be those specified in the `SELECT` clause of the first part of the query.

Reviewer: I think we should add a link to UNION documentation (not sure where that is).

Author: More than that: we should add UNION documentation itself 😅

OQL allows you to specify paths to attributes of associated entities in the `ORDER BY` clause.

{{% alert color="info" %}}
In the example below `Sales.Customer` object for Jim Elk does not have any associated objects of the `Sales.Request` entity. Depending on the database, that object will end up either at the beginning or the end of the ordered result. See [NULL Values Order Behavior](/refguide/ordering-behavior/#null-ordering-behavior) for more information.
{{% /alert %}}

```sql
SELECT LastName
FROM Sales.Customer
ORDER BY Sales.Customer/Sales.Customer_Request/Sales.Request/Number
```

| LastName     |
| ------------ |
| Doe          |
| Moose        |
| Elk          |

### `LIMIT` and `OFFSET` Clauses

With the `LIMIT` and `OFFSET` clauses, you can specify that only a portion of the result of a query is returned.

{{% alert color="info" %}}
Mendix recommends combining `LIMIT` and `OFFSET` clauses with `ORDER BY` because the return order of rows without an `ORDER BY` clause is undefined and can lead to unpredictable output.
{{% /alert %}}

#### Syntax

The syntax is as follows:

```sql
[ LIMIT number ] [ OFFSET number ]
```

#### `LIMIT` clause

> `LIMIT` specifies how many rows must be returned.

Reviewer: Isn't it the maximum number of rows that are returned. It could be less than LIMIT

For example, the following query skips 2 records and then returns 3 records:

```sql
SELECT Brand, City, LocationNumber
FROM Sales.Location
ORDER BY LocationNumber
LIMIT 3
OFFSET 2
```

| Brand  | City      | LocationNumber |
| ------ | -----     | -----          |
| Rekall | Zwolle    | 3              |
| Rekall | Utrecht   | 4              |
| Veidt  | Utrecht   | 5              |

### Subqueries

A subquery is an OQL query nested inside another query. A subquery can contain the same clauses as a regular OQL query. The entities from the outer query can be referred to in a subquery. A subquery can be used in different parts of the query.

#### Subquery in `SELECT` {#subquery-in-select}

 A subquery can be used as a column in the `SELECT` clause. It can refer to other tables and expressions in `FROM`.

```sql
SELECT
	Req/Number AS RequestNumber,
	(
		SELECT COUNT(*)
		FROM Sales.Customer AS Cust
		WHERE Cust/LastName = Req/CustomerName
	) AS CustomerCount
FROM Sales.Request Req
```

| RequestNumber | CustomerCount |
| ------------- | ------------- |
| 1             | 1             |
| 2             | 1             |
| -1            | 0             |

{{% alert color="info" %}}

> If you use a subquery as an expression in `SELECT`, then such subquery should always return a single column, and the number of rows should be at most one. If the subquery returns more than one row or a number of columns different than one, that will lead to an exception during runtime.

Reviewer: Could we just say a single value? Or saying the same thing in a shorter way:
```suggestion
If you use a subquery as an expression in `SELECT`, then such subquery should always return a single row and column. If the subquery returns more than one row or column, that will lead to an exception during runtime.
```

| Brand  | City      | LocationNumber |
| ------ | -----     | -----          |
| Rekall | Zwolle    | 3              |
| Rekall | Utrecht   | 4              |
| Veidt  | Utrecht   | 5              |
| Veidt  | Rotterdam | 6              |

`LIMIT` and `OFFSET` can be combined in one query.

For example, the following query skips 2 records and then returns 3 records:

```sql
SELECT Brand, City, LocationNumber
FROM Sales.Location
ORDER BY LocationNumber
LIMIT 3
OFFSET 2
```

| Brand  | City      | LocationNumber |
| ------ | -----     | -----          |
| Rekall | Zwolle    | 3              |
| Rekall | Utrecht   | 4              |
| Veidt  | Utrecht   | 5              |

## Subqueries

A subquery is an OQL query nested inside another query. A subquery can contain the same clauses as a regular OQL query. The entities from the outer query can be referred to in a subquery. A subquery can be used in different parts of the query.

### Subquery in `SELECT` {#subquery-in-select}

 A subquery can be used as a column in the `SELECT` clause. It can refer to other tables and expressions in `FROM`.

```sql
SELECT
	Req/Number AS RequestNumber,
	(
		SELECT COUNT(*)
		FROM Sales.Customer AS Cust

> 		WHERE Cust/LastName = Req/CustomerName

Reviewer: Will this work, or does COUNT(*) need an AS?

Reviewer 3: `AS` needed.

Reviewer 2: `AS` is not required, same reason as other case with subquery in attribute

Author: In OQL v2, AS should be required

```sql
SELECT FirstName, LastName
FROM Sales.Customer Cust
WHERE
	Cust/LastName IN (
		SELECT CustomerName
		FROM Sales.Request Req
	)
```

| FirstName | LastName |
| --------- | -------- |
| John      | Doe      |
| Jane      | Moose    |

#### Subquery with `EXISTS`

A subquery can be combined with the `EXISTS` keyword. In that case, the expression is true if the subquery returns at least one row. In case of `EXISTS`, the subquery can return any number of rows and any number of columns.

```sql
SELECT FirstName, LastName
FROM Sales.Customer Cust
WHERE
	EXISTS (
		SELECT * FROM Sales.Request Req
		WHERE Req/CustomerName = Cust/LastName
	)
```

| FirstName | LastName |
| --------- | -------- |
| John      | Doe      |
| Jane      | Moose    |

### Subquery in `HAVING` {#subquery-in-having}

Subqueries can be used in a `HAVING` clause the in same way they are used in `WHERE`: as a value or combined with `IN` or `EXISTS` clauses. For other cases, the same limitations apply to the subquery outcome.

Example:

```sql
SELECT COUNT(*) AS LocationCount, SUM(Stock) as CityStock, City AS City
FROM Sales.Location AS Location
GROUP BY City
HAVING
	SUM(Stock) <= (
		SELECT COUNT(*)

> 		FROM Sales.Location

Reviewer: Will this work, or does COUNT(*) need an AS?

Reviewer 3: Same asnwer; `AS` needed.

Reviewer 2: `AS` is not required, `HAVING` and `WHERE` are not affected by changes to mandatory column names. It only applies to the main `SELECT` query and the attributes/columns of it.

| --------------- | ----- | ----- |
| 562949953421521 | John  | Doe   |
| 562949953421683 | Jane  | Doe   |
| 562949953421777 | Jane  | Doe   |
| 562949953421923 | Jane  | Moose |

### Expressions

It is possible to use more complex expressions in `SELECT`. This is explained in detail in [OQL Expressions](/refguide/oql-expressions/).

It is also possible to use a subquery. See [Subquery in `SELECT`](/refguide/oql-clauses/#subquery-in-select) for more details.

### Selecting Attributes over Associations

A unique feature of OQL is the ability to access attributes of associated objects using paths. For example:

```sql
SELECT
	Number AS RequestNumber,
	Sales.Request/Sales.Request_Customer/Sales.Customer/LastName AS CustomerName
FROM Sales.Request
```

It is possible to build paths over multiple associations. Associated entities can be reached from both directions. System associations `System.owner` and `System.changedBy` can also be used in such association paths, assuming they are enabled for the entity. For example:

```sql
SELECT
	LastName AS CustomerName,
	Sales.Customer/Sales.Request_Customer/Sales.Request/System.owner/System.User/Name AS UserName
FROM Sales.Customer
```

> In the case when the association [multiplicity type](/refguide/association-properties/#multiplicity) is one-to-many or many-to-many, every attribute over association will result in multiple results per object. If there are multiple attributes over association in the `FROM` clause, the result will be a cartesian product of associated objects meaning that there will be a row for every combination of associated objects. If you want to avoid that effect, consider rewriting the query using [`JOIN`](/refguide/oql-clauses/#join).

Reviewer: ```suggestion
In the case when the association [multiplicity type](/refguide/association-properties/#multiplicity) is one-to-many or many-to-many, every attribute over association may result in multiple results per object. If there are multiple attributes over association in the `FROM` clause, the result will be a cartesian product of associated objects meaning that there will be a row for every combination of associated objects. If you want to avoid that effect, consider rewriting the query using [`JOIN`](/refguide/oql-clauses/#join).
```

> For example, when a `Sales.Customer` with `LastName` "Doe" has two associated objects in `Sales.Request` with `Number` 1 and 2, the following query will return 4 rows:

Reviewer: ```suggestion
For example, when a `Sales.Customer` with `LastName` "Doe" has two associated  `Sales.Request` objects with a `Number` attribute with values 1 and 2, the following query will return 4 rows:
```

| Rekall | Utrecht   | 4              |
| Veidt  | Utrecht   | 5              |

> ## `UNION` Clause {#oql-union}

Author: We should add an example of how `UNION` works with association columns

The clause takes multiple select queries and combines their results into a single result set.
The resulting set by default only includes distinct rows. The `ALL` keyword can be used to include all rows. Rows are considered distinct if the combination of the column values is distinct from all other rows. Comparison logic of values is the same as the `DISTINCT` keyword of a `SELECT` clause.

> All select queries must define the same number of columns in the same order and their datatypes should match. The resulting column names will those used in the very first `SELECT` clause.

Author: There is probably a word missing in `The resulting column names will those used`

Reviewer: ```suggestion
All select queries must define the same number of columns in the same order and their datatypes should match. The resulting column names will be those used in the very first `SELECT` clause.
```
